examples/dao/assets/dao-app-approval.py

```diff
@@ -0,0 +1,471 @@
+    ])
+
+    # p_<proposal> is a concatenation of p_ with the proposal address to avoid some weird attacks.
+    byte_p_proposal = Concat(Bytes("p_"), Txn.accounts[1])
+    p_proposal = App.localGetEx(Int(0), Int(0), byte_p_proposal)  # value = proposal.id when a user voted or 0
+    yes = Bytes("yes")
+    no = Bytes("no")
+    abstain = Bytes("abstain")
+
+    # Register user votes in proposal_lsig by saving Sender.p_<proposal>. Arguments:
+    # * proposal : lsig account address with the proposal record (provided as the first external account).
+    # * vote (bytes): abstain, yes, no
+    register_vote = Seq([
+        p_proposal,
+        Assert(
+            And(
+                Global.group_size() == Int(1),
+                # voting_start <= now <= voting_end
+                # voting_start <= Global.latest_timestamp(),
+                # Global.latest_timestamp() <= voting_end,
+                # Sender.deposit >= 0 (i.e user "deposited" his votes using deposit_vote_token)
+                App.localGet(Int(0), Bytes("deposit")) > Int(0)
+            )
+        ),
+        If(
+            p_proposal.hasValue() == Int(0),
+            # If Sender.p_<proposal> is not set then set p_<proposal> := proposal.id
+            App.localPut(Int(0), byte_p_proposal, proposal_id),
+            # if Sender.p_<proposal> != proposal.id then overwrite by setting the new proposal.id, fail otherwise
+            If(p_proposal.value() != proposal_id,
+                App.localPut(Int(0), byte_p_proposal, proposal_id),
+                Err()),
+        ),
+        # record vote in proposal_lsig local state (proposal.<counter> += Sender.deposit)
+        Cond(
+            [Gtxn[0].application_args[1] == yes, App.localPut(Int(1), yes, App.localGet(Int(1), yes) + deposit)],
+            [Gtxn[0].application_args[1] == no, App.localPut(Int(1), no, App.localGet(Int(1), no) + deposit)],
+            [Gtxn[0].application_args[1] == abstain, App.localPut(Int(1), abstain, App.localGet(Int(1), abstain) + deposit)]
+        ),
+        # Update Sender.deposit_lock := max(Sender.deposit_lock, proposal.voting_end)
+        If(
+            App.localGet(Int(0), Bytes("deposit_lock")) <= voting_end,
+            App.localPut(Int(0), Bytes("deposit_lock"), voting_end)
+        ),
+        Return(Int(1))
+    ])
+
+    # Clears Sender local state by removing a record of vote cast from a not active proposal. Args:
+    # * proposal : lsig account address with the proposal record (provided as the first external account).
+    clear_vote_record = Seq([
+        p_proposal,
+        scratch_proposal_active(Int(1)),
+        Assert(Global.group_size() == Int(1)),
+        # fail if proposal is active (can’t remove proposal record of an active proposal)
+        If(
+            p_proposal.hasValue() == Int(1),
+            If(
+                And(
+                    p_proposal.value() == proposal_id,
+                    scratchvar_proposal_active.load() == Int(1)
```

Comment:
clearing a vote record should be only possible when the proposal status `== 3` (is active).
`1` means that voting is over and it passed.

Reply:
Could you please check the spec: [https://paper.dropbox.com/doc/Algo-Builder-DAO--BSmq7oxrGfMChl9VhEhjwJO4Ag-ncLdytuFa7EJrRerIASSl](https://paper.dropbox.com/doc/Algo-Builder-DAO--BSmq7oxrGfMChl9VhEhjwJO4Ag-ncLdytuFa7EJrRerIASSl) ?

`scratch_proposal_active()` is implemented as per spec, where `1` means proposal is still in voting.

Reply:
ah, ok. Please add a comment to the code.